### PR TITLE
First implementation of a Stack

### PR DESCRIPTION
**Pull request by ClementWalter**
Things to discuss before merge:
- relevance of `Bytes32` struct instead of used `Uint256`
- possibility to remove the size `StackStruct` and consequently keep only a `Uint256*`
   - did not see how to check that the stack was empty or full without the `size`

### CODE REVIEW

**Comment by HrikB** [comment]
If the standard library already provides a `Uint256` then I see no reason to make our own `Bytes32` struct. However I think it would be nice for the sake of being explicit to alias the `Uint256` to `Bytes32`. Although not necessary.

File: src/onlydust/imhotep/stack.cairo [hunk 1]
```diff
@@ -0,0 +1,48 @@
+%lang starknet
+
+from starkware.cairo.common.uint256 import Uint256
+from starkware.cairo.common.math_cmp import is_not_zero
+from starkware.cairo.common.bool import FALSE
+from starkware.cairo.common.math import assert_lt_felt
+
+struct StackStruct {
```
Comment (Bernardstanislas) [inline src/onlydust/imhotep/stack.cairo:9]: Maybe we can simply name it `Stack`?

File: tests/stack/test_push.cairo [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    let slots: Uint256* = alloc();
+    let max_stack_size = Stack.MAX_STACK_SIZE;
+    let stack = StackStruct(size=0, slots=slots);
+    _stack_fixture_loop{stack=stack, target_stack_size=max_stack_size}(0);
+    let slot = Uint256(max_stack_size + 1, 0);
+    Stack.push{stack=stack}(slot);
+    return ();
+}
+
+func _stack_fixture_loop{
```
Comment (Bernardstanislas) [inline tests/stack/test_push.cairo]:
```diff
- func _stack_fixture_loop{
+ func _full_stack_fixture_loop{
```
?

File: src/onlydust/imhotep/stack.cairo (hunk 1 above)
Comment (ClementWalter) [reply src/onlydust/imhotep/stack.cairo:9]: but then the `stack` variable would have to be named differently. It assumed that the namespace is somehow the pendant of a class name

File: tests/stack/test_push.cairo (hunk 2 above)
Comment (ClementWalter) [reply tests/stack/test_push.cairo]: "full" up to "target_stack_size" so might not be full indeed

**Comment by ClementWalter** [comment]
no idea on how to remove the `size`? I really think that there should be a way to check what's behind the pointer because if you try to `[wrong memory address]` then it fails

**Comment by Bernardstanislas** [comment]
> no idea on how to remove the `size`? I really think that there should be a way to check what's behind the pointer because if you try to `[wrong memory address]` then it fails

It don't think it exists in Cairo, arrays are simply pointers with a size. Look at this https://www.cairo-lang.org/docs/hello_starknet/more_features.html#array-arguments-in-calldata

File: tests/stack/test_push.cairo (hunk 2 above)
Comment (Bernardstanislas) [reply tests/stack/test_push.cairo]: oh that's true I didn't see

File: src/onlydust/imhotep/stack.cairo (hunk 1 above)
Comment (Bernardstanislas) [reply src/onlydust/imhotep/stack.cairo:9]: ok. I don't have a solution in mind

File: tests/stack/test_push.cairo (hunk 2 above)
Comment (Bernardstanislas) [reply tests/stack/test_push.cairo]: That's true, didn't see the target size argument

**Review by Bernardstanislas — APPROVED** [review]

**Comment by ClementWalter** [comment]
Resolve #15

**Pull request merged**